packages/edit-post/src/components/sidebar/settings-header/index.js

```diff
@@ -3,54 +3,84 @@
  */
 import { __ } from '@wordpress/i18n';
 import { withDispatch } from '@wordpress/data';
+import { TabPanel } from '@wordpress/components';
+import { useState, useEffect } from '@wordpress/element';
 
 /**
  * Internal dependencies
  */
 import SidebarHeader from '../sidebar-header';
 
 const SettingsHeader = ( { openDocumentSettings, openBlockSettings, sidebarName } ) => {
-	const blockLabel = __( 'Block' );
-	const [ documentAriaLabel, documentActiveClass ] = sidebarName === 'edit-post/document' ?
+	const [ documentAriaLabel, documentSettingsActive ] = sidebarName === 'edit-post/document' ?
 		// translators: ARIA label for the Document sidebar tab, selected.
-		[ __( 'Document (selected)' ), 'is-active' ] :
+		[ __( 'Document (selected)' ), true ] :
 		// translators: ARIA label for the Document sidebar tab, not selected.
-		[ __( 'Document' ), '' ];
+		[ __( 'Document' ), false ];
 
-	const [ blockAriaLabel, blockActiveClass ] = sidebarName === 'edit-post/block' ?
+	const [ blockAriaLabel, blockSettingsActive ] = sidebarName === 'edit-post/block' ?
 		// translators: ARIA label for the Settings Sidebar tab, selected.
-		[ __( 'Block (selected)' ), 'is-active' ] :
+		[ __( 'Block (selected)' ), true ] :
 		// translators: ARIA label for the Settings Sidebar tab, not selected.
-		[ __( 'Block' ), '' ];
+		[ __( 'Block' ), false ];
+
+	const documentSettingsName = 'edit-post-settings__panel-tab-document';
+	const blockSettingsName = 'edit-post-settings__panel-tab-block';
+
+	const [ selectedTabName, setSelectedTabName ] = useState( documentSettingsName );
+
+	const tabs = [
+		{
+			name: documentSettingsName,
+			className: 'edit-post-sidebar__panel-tab',
+			title: __( 'Document' ),
+			ariaLabel: `${ documentAriaLabel }`,
+		},
+		{
+			name: blockSettingsName,
+			className: 'edit-post-sidebar__panel-tab',
+			title: __( 'Block' ),
+			ariaLabel: `${ blockAriaLabel }`,
+		},
+	];
+
+	const onSelect = ( tabName ) => {
+		setSelectedTabName( tabName );
+		if ( tabName === documentSettingsName ) {
+			openDocumentSettings();
+		} else if ( tabName === blockSettingsName ) {
+			openBlockSettings();
+		}
+	};
+
+	useEffect(
```

Comment:
Did you consider adding `onSelect` to the objects passed to `tabs` instead? It might simplify the solution as you could leave the whole logic which reasons which tab is selected to `TabPanel` component.

Reply:
I had considered it, but avoided that route as I was originally trying to minimize changes to `TabPanel`. I agree adding an `onSelect` method to the tab objects would simplify the solution. (We would still need to support the `onSelect` component handler for backwards compatibility)

The issue that led me to refactor `TabPanel` was the need to control the selected tab from the parent component without triggering a re-render of the entire component, causing the active tab to lose focus.

Reply:
> The issue that led me to refactor `TabPanel` was the need to control the selected tab from the parent component without triggering a re-render of the entire component, causing the active tab to lose focus.

Shouldn't the component ensure that there is never focus loss? 3rd party developers shouldn't care about it, it should be baked in into such component.

Reply:
I will think about that, thank you!

Reply:
I believe I was able to simplify the logic, please let me know if [these changes](https://github.com/WordPress/gutenberg/pull/16663/commits/068604458b3fe690d7d041011850056104a0ec30) address your feedback @gziolo, thanks.